Implement e2e flow with pre-existing DKG rituals for tDec

piotr-roslaniec opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [x] Feature
- [ ] Documentation
- [ ] Refactor
- [ ] Other

**Required reviews:** 
> How many reviews does the PR author need?
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
- Implements e2e flow for using DKG in tDec

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

piotr-roslaniec commented:
Not ready for review just yet, but looking for comments on API design with the `CbdStrategy` and `PreStrategy` separation. Naming and other decisions are up for debate, this is just a rough sketch of what I'm going for.

CC: @derekpierre @KPrasch @theref @manumonti

derekpierre commented on src/sdk/strategy/cbd-strategy.ts:
```diff
@@ -0,0 +1,132 @@
+
+    const decrypter = new CbdTDecDecrypter(
+      this.cohort.configuration.porterUri,
+      dkgRitual.dkgPublicKey
```
Why does the decrypter need the `dkgPublicKey`?

piotr-roslaniec replied:
It doesn't, that's a refactoring artifact. I'm going to remove it.

derekpierre commented:
> Not ready for review just yet, but looking for comments on API design with the `CbdStrategy` and `PreStrategy` separation. Naming and other decisions are up for debate, this is just a rough sketch of what I'm going for.
> 
> CC: @derekpierre @KPrasch @theref @manumonti

Perhaps a walkthough session of your initial thoughts here in the PR could be good? wdyt? It'll be good (at least for me 😅 ) to better visualize/discuss it.

piotr-roslaniec commented on src/characters/cbd-universal-bob.ts:
```diff
@@ -0,0 +1,125 @@
+  ): Promise<DecryptionSharePrecomputed[] | DecryptionShareSimple[]> {
+    const contextStr = await conditionSet.buildContext(provider).toJson();
+
+    // TODO: Move ThresholdDecryptionRequest creation and parsing to Porter?
```
Revisit after integrating e2e encryption

piotr-roslaniec commented on src/characters/cbd-universal-bob.ts:
```diff
@@ -0,0 +1,125 @@
+      ciphertext
+    );
+
+    // TODO: Replace with a factory method
```
Will be superseded in a future release of `ferveo`

theref commented on src/characters/pre-universal-bob.ts:
```diff
@@ -8,23 +8,23 @@ import {
 };
 
-export class tDecDecrypter {
+export class PreTDecDecrypter {
```
Do we really want to keep this?

piotr-roslaniec replied:
I don't think we have a plan for PRE phase-out yet

theref replied:
My vote would be to rip the band aid off and just get rid asap. It will annoy some adopters for sure, but we can support them switching over.

KPrasch replied:
Current development goals do not include phasing out PRE. We'll continue to support it as part of the Threshold Access Control product for the foreseeable future.

theref replied:
We're not saying phase out PRE, but PRE based tDec

KPrasch replied:
Ah, got it! That was unclear from the above comments.  Well in that case let's make a plan. 😁

theref commented on src/sdk/strategy/cbd-strategy.ts:
```diff
@@ -0,0 +1,126 @@
+  conditionSet?: ConditionSet;
+};
+
+type DeployedStrategyJSON = {
```
At what point do we think these objects will be deprecated and removed?

theref replied:
that's a very good point. Separate them out, finalise the CBD design, remove legacy code 🚀

cygnusv commented on src/sdk/strategy/cbd-strategy.ts:
```diff
@@ -0,0 +1,126 @@
+import { ethers } from 'ethers';
+
+import { CbdTDecDecrypter } from '../../characters/cbd-universal-bob';
```
Reconsider the name `cbd-universal-bob.ts`. At the very least, we should drop the `universal` (which made sense in the context of fake, PRE-based TDEC), although I'm much in favour in getting rid of character names, so perhaps something like `cbd-consumer` or `cbd-recipient`.

theref commented on src/index.ts:
```diff
@@ -2,7 +2,7 @@
 export { Bob, RemoteBob } from './characters/bob';
 export { Enrico } from './characters/enrico';
-export { PreTDecDecrypter } from './characters/pre-universal-bob';
+export { PreTDecDecrypter } from './characters/pre-recipient';
```
🚀

piotr-roslaniec commented:
Rebased over parent branch. Added some changes on the back of the rebase, mostly renaming and extra tests.
If there are no objections, I'd like to merge this.

piotr-roslaniec commented on src/sdk/strategy/cbd-strategy.ts:
```diff
@@ -0,0 +1,149 @@
+import { fromJSON, objectEquals, toJSON } from '../../utils';
+import { Cohort, CohortJSON } from '../cohort';
+
+type StrategyJSON = {
```
Renames
```suggestion
type CbdStrategyJSON = {
```

KPrasch commented:
LGTM
